File: src/python/filter_ics/naming.py
"""Output-name sanitization for feed files (see contracts/publishing.md G4)."""

from __future__ import annotations

import re

from .allowlist import FeedError

# A safe single path segment: letters, digits, dot, dash, underscore.
_SAFE = re.compile(r"^[A-Za-z0-9._-]+$")


def sanitize_output_name(file_name: str) -> str:
    """Return a safe base name (no extension) for the feed's `.ics`.

    Strips a trailing ``.ics`` (case-insensitive) once, then validates the result is a
    single safe path segment — no separators, no ``.``/``..``, no leading dot. Raises
    :class:`FeedError` for empty/unsafe values (prevents path traversal).
    """
    name = file_name.strip()
    if name.lower().endswith(".ics"):
        name = name[:-4]
    if (
        not name
        or name in (".", "..")
        or "/" in name
        or "\\" in name
        or name.startswith(".")
        or not _SAFE.match(name)
    ):
        raise FeedError(f"unsafe or empty fileName: {file_name!r}")
    return name


def output_filename(output_name: str) -> str:
    """The published file name for a sanitized output name."""
    return f"{output_name}.ics"

File: src/dotnet/FilterIcs/Naming.cs
using System.Text.RegularExpressions;

namespace FilterIcs;

/// <summary>Output-name sanitization for feed files (contracts/publishing.md G4).</summary>
public static partial class Naming
{
    [GeneratedRegex("^[A-Za-z0-9._-]+$")]
    private static partial Regex SafeSegment();

    /// <summary>
    /// Return a safe base name (no extension). Strips a trailing ".ics" once, then validates a
    /// single safe path segment (no separators, no '.'/'..', no leading dot). Throws on unsafe/empty.
    /// </summary>
    public static string SanitizeOutputName(string fileName)
    {
        string name = fileName.Trim();
        if (name.EndsWith(".ics", StringComparison.OrdinalIgnoreCase))
            name = name[..^4];

        if (string.IsNullOrEmpty(name)
            || name is "." or ".."
            || name.Contains('/')
            || name.Contains('\\')
            || name.StartsWith('.')
            || !SafeSegment().IsMatch(name))
        {
            throw new FeedException($"unsafe or empty fileName: '{fileName}'");
        }
        return name;
    }

    public static string OutputFileName(string outputName) => outputName + ".ics";
}
